modules/tfm/tfm/boards/common/attest_hal.c

```diff
@@ -15,34 +15,195 @@
+	uint32_t val32 = *(uint32_t *)target_addr;
+	__DSB(); /* Because of nRF9160 Erratum 7 */
+
+	return (top_half ? (val32 >> 16) : val32) & 0x0000FFFF;
+}
+
+/** Function for writing a half-word (2 byte value) to OTP.
+ *
+ * \details The flash in OTP supports only aligned 4 byte write operations.
+ *          This function writes to the encompassing 4 byte word, masking the
+ *          other half-word with 0xFFFF so it is left untouched.
+ *
+ * \param[in]  ptr  Address to write to.
+ * \param[in]  val  Value to write into @p ptr.
+ */
+static void write_halfword(const uint16_t *ptr, uint16_t val)
+{
+	bool top_half = (uint32_t)ptr % 4; /* Addr not div by 4 */
+	uint32_t target_addr = (uint32_t)ptr & ~3; /* Floor address */
+
+	uint32_t val32 = (uint32_t)val | 0xFFFF0000;
+	uint32_t val32_shifted = ((uint32_t)val << 16) | 0x0000FFFF;
+
+	nrfx_nvmc_word_write(target_addr, top_half ? val32_shifted : val32);
+}
+
+
+/* Can later be replaced by a bl_storage call */
+int read_lcs_from_otp(lcs_t *lcs)
+{
+	uint16_t left_assembly     = read_halfword(&p_bl_storage_data->lcs.left_assembly);
+	uint16_t left_provisioning = read_halfword(&p_bl_storage_data->lcs.left_provisioning);
+	uint16_t left_secure       = read_halfword(&p_bl_storage_data->lcs.left_secure);
+
+	if (left_assembly == STATE_NOT_LEFT) {
+		*lcs = ASSEMBLY;
+	} else if (left_assembly == STATE_LEFT && left_provisioning == STATE_NOT_LEFT) {
+		*lcs = PROVISION;
+	} else if (left_provisioning == STATE_LEFT && left_secure == STATE_NOT_LEFT) {
+		*lcs = SECURE;
+	} else if (left_provisioning == STATE_LEFT) {
+		*lcs = DECOMMISSIONED;
+	} else {
+		/* To reach this the OTP must be corrupted or reading failed */
+		return -EREADLCS;
+	}
+
+	return 0;
+}
+
+int update_lcs_in_otp(lcs_t next_lcs)
+{
+	int err;
+	lcs_t current_lcs = 0;
+
+	if(next_lcs == UNKNOWN){
+		return -EINVALIDLCS;
+	}
+
+	err = read_lcs_from_otp(&current_lcs);
+	if (err != 0) {
+		return err;
+	}
+
+	if (next_lcs < current_lcs) {
+		/* Is is only possible to transition into a higher state */
+		return -EINVALIDLCS;
+	}
+
+	if (next_lcs == current_lcs) {
+		/* The same LCS is a valid argument, but nothing to do so return success */
+		return 0;
+	}
+
+	/* As the device starts in ASSEMBLY, it is not possible to write it */
+	if (current_lcs == ASSEMBLY && next_lcs == PROVISION) {
+		write_halfword(&(p_bl_storage_data->lcs.left_assembly), STATE_LEFT);
+		return 0;
+	}
+
+	if (current_lcs == PROVISION && next_lcs == SECURE) {
+		write_halfword(&(p_bl_storage_data->lcs.left_provisioning), STATE_LEFT);
+		return 0;
+	}
+
+	if (current_lcs == SECURE && next_lcs == DECOMMISSIONED) {
+		write_halfword(&(p_bl_storage_data->lcs.left_secure), STATE_LEFT);
+		return 0;
+	}
+
+	/* This will be the case if any invalid transition is tried */
+	return -EINVALIDLCS;
+}
+
+static void read_implementation_id_from_otp(uint8_t * buf)
+{
+	otp_copy32(buf, (uint32_t *)&p_bl_storage_data->implementation_id, 64);
```

Comment:
Magic constant.
```suggestion
	otp_copy32(buf, (uint32_t *)&p_bl_storage_data->implementation_id, sizeof(p_bl_storage_data->implementation_id));
```

Reply:
sizeof would imply to the reader that this could be any size.

I think 32 would communicate more clearly the reality.

Reply:
added define